File: examples/pubsub_example.js
const amqp = require('amqplib');

// Connection URL
const url = 'amqp://guest:guest@localhost:5672';

async function publisher() {
    try {
        // Establish connection
        const connection = await amqp.connect(url);
        const channel = await connection.createChannel();
        
        // Create fanout exchange
        const exchange = 'logs';
        await channel.assertExchange(exchange, 'fanout', { durable: false });
        
        // Publish messages
        for (let i = 0; i < 5; i++) {
            const message = `Broadcast message ${i}`;
            channel.publish(exchange, '', Buffer.from(message));
            console.log(` [x] Sent ${message}`);
        }
        
        // Close connection
        setTimeout(() => {
            connection.close();
        }, 500);
    } catch (error) {
        console.error('Error:', error);
    }
}

async function subscriber() {
    try {
        // Establish connection
        const connection = await amqp.connect(url);
        const channel = await connection.createChannel();
        
        // Create fanout exchange
        const exchange = 'logs';
        await channel.assertExchange(exchange, 'fanout', { durable: false });
        
        // Create temporary queue
        const q = await channel.assertQueue('', { exclusive: true });
        
        // Bind queue to exchange
        await channel.bindQueue(q.queue, exchange, '');
        
        console.log(' [*] Waiting for messages. To exit press CTRL+C');
        
        // Consume messages
        channel.consume(q.queue, (msg) => {
            if (msg !== null) {
                console.log(` [x] Received ${msg.content.toString()}`);
            }
        }, { noAck: true });
    } catch (error) {
        console.error('Error:', error);
    }
}

// Check command line arguments
const args = process.argv.slice(2);
if (args.length > 0 && args[0] === 'publish') {
    publisher();
} else {
    subscriber();
}

File: examples/pubsub_example.py
import pika
import sys
import time

def publisher():
    connection = pika.BlockingConnection(
        pika.ConnectionParameters(host='localhost'))
    channel = connection.channel()

    # Create fanout exchange
    channel.exchange_declare(exchange='logs', exchange_type='fanout')

    # Publish messages
    for i in range(5):
        message = f'Broadcast message {i}'
        channel.basic_publish(
            exchange='logs',
            routing_key='',
            body=message
        )
        print(f" [x] Sent {message}")
    
    connection.close()

def subscriber():
    connection = pika.BlockingConnection(
        pika.ConnectionParameters(host='localhost'))
    channel = connection.channel()

    # Create fanout exchange
    channel.exchange_declare(exchange='logs', exchange_type='fanout')

    # Create temporary queue
    result = channel.queue_declare(queue='', exclusive=True)
    queue_name = result.method.queue

    # Bind queue to exchange
    channel.queue_bind(exchange='logs', queue=queue_name)

    def callback(ch, method, properties, body):
        print(f" [x] Received {body.decode()}")
        time.sleep(1)

    channel.basic_consume(
        queue=queue_name,
        on_message_callback=callback,
        auto_ack=True
    )

    print(' [*] Waiting for messages. To exit press CTRL+C')
    channel.start_consuming()

if __name__ == '__main__':
    try:
        if len(sys.argv) > 1 and sys.argv[1] == 'publish':
            publisher()
        else:
            subscriber()
    except KeyboardInterrupt:
        print('Interrupted')
